BEIMA.Backend.Test/MongoService/BaseDeviceTest.cs

```diff
@@ -0,0 +1,207 @@
+            Assert.That((string)fields.GetElement(validKey1).Value, Is.EqualTo(validValue1));
+            Assert.That((bool)fields.GetElement(validKey2).Value, Is.EqualTo(validValue2));
+            Assert.That((int)fields.GetElement(validKey3).Value, Is.EqualTo(validValue3));
+        }
+
+        [Test]
+        public void BaseDeviceNotInstantiated_InstantiateUsingObjectInitializer_BaseDeviceInstantiatedWithCorrectValues()
+        {
+            var device = new BaseDevice
+            {
+                Id = validObjId,
+                DeviceTypeId = validDeviceTypeId,
+                DeviceTag = validDeviceTag,
+                Manufacturer = validManufacturer,
+                ModelNum = validModelNum,
+                SerialNum = validSerialNum,
+                YearManufactured = validYearManufactured,
+                Notes = validNotes
+            };
+            Assert.That(device.Id, Is.EqualTo(validObjId));
+            Assert.That(device.DeviceTypeId, Is.EqualTo(validDeviceTypeId));
+            Assert.That(device.DeviceTag, Is.EqualTo(validDeviceTag));
+            Assert.That(device.Manufacturer, Is.EqualTo(validManufacturer));
+            Assert.That(device.ModelNum, Is.EqualTo(validModelNum));
+            Assert.That(device.SerialNum, Is.EqualTo(validSerialNum));
+            Assert.That(device.YearManufactured, Is.EqualTo(validYearManufactured));
+            Assert.That(device.Notes, Is.EqualTo(validNotes));
+
+            device.SetLastModified(validDate, validUser);
+            var lastModified = device.LastModified;
+            Assert.That((DateTime)lastModified.GetElement("date").Value, Is.EqualTo(validDate).Within(5).Seconds);
+            Assert.That((string)lastModified.GetElement("user").Value, Is.EqualTo(validUser));
+
+            device.SetLocation(validBuildingId, validLocationNotes, validLatitude, validLongitude);
+            var location = device.Location;
+            Assert.That((ObjectId)location.GetElement("buildingId").Value, Is.EqualTo(validBuildingId));
+            Assert.That((string)location.GetElement("notes").Value, Is.EqualTo(validLocationNotes));
+            Assert.That((string)location.GetElement("latitude").Value, Is.EqualTo(validLatitude));
+            Assert.That((string)location.GetElement("longitude").Value, Is.EqualTo(validLongitude));
+
+            device.AddField(validKey1, validValue1);
+            device.AddField(validKey2, validValue2);
+            device.AddField(validKey3, validValue3);
+            var fields = device.Fields;
+            Assert.That((string)fields.GetElement(validKey1).Value, Is.EqualTo(validValue1));
+            Assert.That((bool)fields.GetElement(validKey2).Value, Is.EqualTo(validValue2));
+            Assert.That((int)fields.GetElement(validKey3).Value, Is.EqualTo(validValue3));
+        }
+
+        [Test]
+        public void BaseDeviceNotInstantiated_InstantiateUsingObjectInitializerWithAllNullStringValues_GetBsonDocumentThrowsException()
```

Comment:
If you want, you can set up various test cases using the same method. You would use something like:
`[TestCase(1, "2", null)]
[TestCase(3, "4", "5"]
public void MyTest(int number, string name, string value)`
Although, this would probably need a ridiculous number of method parameters .

Reply:
Since I've already written it this way, and also the number of method parameters is so large, I will keep it like this. Going forward, I will try and make use of TestCases.